components/brave_sync/time_limited_words.cc

```diff
@@ -0,0 +1,209 @@
+}
+
+std::string TimeLimitedWords::GenerateForDate(const std::string& pure_words,
+                                              const Time& not_after) {
+  int days_since_words_v2_epoch =
+      GetRoundedDaysDiff(GetWordsV2Epoch(), not_after);
+
+  if (days_since_words_v2_epoch < 0) {
+    // Something goes bad, requested |not_after| is even before sync v2 epoch
+    return std::string();
+  }
+
+  std::string last_word = GetWordByIndex(days_since_words_v2_epoch);
+
+  std::string time_limited_code = base::StrCat({pure_words, " ", last_word});
+  return time_limited_code;
+}
+
+WordsValidationStatus TimeLimitedWords::Validate(
+    const std::string& time_limited_words,
+    std::string* pure_words) {
+  CHECK_NE(pure_words, nullptr);
+  *pure_words = std::string();
+
+  static constexpr size_t kPureWordsCount = 24u;
+  static constexpr size_t kWordsV2Count = 25u;
+
+  auto now = Time::Now();
+
+  std::vector<std::string> words = base::SplitString(
+      time_limited_words, " ", base::WhitespaceHandling::TRIM_WHITESPACE,
+      base::SplitResult::SPLIT_WANT_NONEMPTY);
+
+  size_t num_words = words.size();
+
+  if (num_words == kPureWordsCount) {
+    if (now < GetWordsV1SunsetDay()) {
+      std::string recombined_pure_words = base::JoinString(
+          base::span<std::string>(words.begin(), kPureWordsCount), " ");
+      if (crypto::IsPassphraseValid(recombined_pure_words)) {
+        *pure_words = recombined_pure_words;
+        return WordsValidationStatus::kValid;
+      } else {
+        return WordsValidationStatus::kNotValidPureWords;
+      }
+    } else {
+      return WordsValidationStatus::kVersionDeprecated;
+    }
+  } else if (num_words == kWordsV2Count) {
+    std::string recombined_pure_words = base::JoinString(
+        base::span<std::string>(words.begin(), kPureWordsCount), " ");
+    if (crypto::IsPassphraseValid(recombined_pure_words)) {
+      int days_actual =
+          GetRoundedDaysDiff(GetWordsV2Epoch(), now) % BIP39_WORDLIST_LEN;
+
+      int days_encoded = GetIndexByWord(words[kWordsV2Count - 1]);
+      DCHECK(days_encoded < BIP39_WORDLIST_LEN);
+
+      int days_abs_diff = std::abs(days_actual - days_encoded);
+      if (days_abs_diff <= 1) {
+        *pure_words = recombined_pure_words;
+        return WordsValidationStatus::kValid;
+      } else if (days_actual > days_encoded) {
+        return WordsValidationStatus::kExpired;
+      } else if (days_encoded > days_actual) {
+        return WordsValidationStatus::kValidForTooLong;
+      }
+    } else {
+      return WordsValidationStatus::kNotValidPureWords;
+    }
+  } else {
+    return WordsValidationStatus::kWrongWordsNumber;
+  }
+
+  NOTREACHED();
+  return WordsValidationStatus::kNotValidPureWords;
+}
+
+TimeLimitedWords::PureWordsWithStatus TimeLimitedWords::Parse(
+    const std::string& time_limited_words) {
+  PureWordsWithStatus ret;
+  std::string pure_words;
+  ret.status = Validate(time_limited_words, &pure_words);
+
```

Comment:
please initiate ret.pure_words as `absl::nullopt`

Reply:
Fixed at https://github.com/brave/brave-core/pull/12300/commits/9efddfb1fd3e969ae45617a9ba9f92c023e06557 , but I was sure it is handled by constructor

Reply:
It's safer to initiate it because we can't be sure that constructor behavior won't change